statsmodels/sysreg/sysmodel.py

```diff
@@ -0,0 +1,622 @@
+        constraints on parameters. See Notes.
+    restrict_vect : ndarray, (M, 1)
+        The RHS restriction vector. See Notes.
+
+    Attributes
+    ----------
+    cholsigmainv : ndarray
+        The transpose of the Cholesky decomposition of the pseudoinverse of
+        the contemporaneous covariance matrix.
+    wendog : ndarray, (neqs*nobs, 1)
+        Endogenous variables whitened by cholsigmainv and stacked into a single
+        column.
+    wexog : ndarray
+        Whitened exogenous variables sp_exog.
+    pinv_wexog : ndarray
+        Moore-Penrose pseudoinverse of `wexog`.
+    normalized_cov_params : array
+
+    Notes
+    -----
+    Linear restrictions on parameters are specified with the following equation:
+        restrict_matrix * params = restrict_vect
+    '''
+
+    def __init__(self, sys, sigma=None, restrict_matrix=None, restrict_vect=None):
+        super(SysGLS, self).__init__(sys)
+
+        ## Handle sigma
+        if sigma is None:
+            self.sigma = np.diag(np.ones(self.neqs))
+        # sigma = scalar
+        elif sigma.shape == ():
+            self.sigma = np.diag(np.ones(self.neqs)*sigma)
+        # sigma = 1d vector
+        elif (sigma.ndim == 1) and sigma.size == self.neqs:
+            self.sigma = np.diag(sigma)
+        # sigma = GxG matrix
+        elif sigma.shape == (self.neqs,self.neqs):
+            self.sigma = sigma
+        else:
+            raise ValueError("sigma is not correctly specified")
+
+        ## Handle restrictions
+        self.isrestricted = not (restrict_matrix is None and restrict_vect is None)
+        if self.isrestricted:
+            if not(restrict_vect.shape[0] == restrict_matrix.shape[0]):
+                raise ValueError('restrict_vect and restrict_matrix must have the \
+                    same number of rows')
+            self.nconstraints = restrict_vect.shape[0]
+            if self.nconstraints >= self.k_exog_all:
+                raise ValueError('total number of regressors must be greater than \
+                    the number of constraints')
+            self.restrict_matrix = restrict_matrix
+            self.restrict_vect = restrict_vect
+
+        self.initialize()
+
+    def initialize(self):
+        self.cholsigmainv = np.linalg.cholesky(np.linalg.pinv(self.sigma)).T
+        self.wexog = self.whiten(self.sp_exog)
+        self.wendog = self.whiten(self.endog.T.reshape(-1,1))
+        self.pinv_wexog = np.linalg.pinv(self.wexog)
+
+        if self.isrestricted:
+            rwendog = np.zeros((self.k_exog_all + self.nconstraints,))
+            rwendog[:self.k_exog_all] = np.squeeze(np.dot(self.wexog.T,self.wendog))
+            rwendog[self.k_exog_all:] = self.restrict_vect
+            self.rwendog = rwendog
+
+            rwexog = np.zeros((self.k_exog_all + self.nconstraints,
+                self.k_exog_all + self.nconstraints))
+            rwexog[:self.k_exog_all, :self.k_exog_all] = np.dot(self.wexog.T,
```

Comment:
this looks like a misnomer, it is the cross-product matrix extended with constraints.